File: src/main/java/com/github/pauldambra/moduluschecker/chain/gates/SecondCheckRequiredGate.java
package com.github.pauldambra.moduluschecker.chain.gates;

import com.github.pauldambra.moduluschecker.ModulusCheckParams;
import com.github.pauldambra.moduluschecker.chain.ModulusChainLink;
import com.github.pauldambra.moduluschecker.chain.ModulusResult;
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow;

import java.util.Optional;

public class SecondCheckRequiredGate implements ModulusChainLink {

    private final ExceptionTwoGate next;

    public SecondCheckRequiredGate(ExceptionTwoGate next) {
        this.next = next;
    }

    @Override
    public ModulusResult check(ModulusCheckParams params) {

        if (exceptionRequiresSecondCheck(params)) {
            return next.check(params);
        }

        if (params.secondWeightRow.isPresent()) {
            return next.check(params);
        }

        return params.modulusResult.orElse(ModulusResult.PASSES);
    }

    private boolean exceptionRequiresSecondCheck(ModulusCheckParams params) {
        return rowExceptionRequiresSecondCheck(params.firstWeightRow)
                || rowExceptionRequiresSecondCheck(params.secondWeightRow);
    }

    private boolean rowExceptionRequiresSecondCheck(Optional<WeightRow> row) {
        return row.isPresent()
                && ModulusResult.exceptionsThatRequireSecondCheck.contains(
                    row.get().exception.orElse(-1));
    }
}

File: src/main/java/com/github/pauldambra/moduluschecker/chain/gates/SecondCheckRequiredGate.kt
package com.github.pauldambra.moduluschecker.chain.gates

import com.github.pauldambra.moduluschecker.ModulusCheckParams
import com.github.pauldambra.moduluschecker.chain.ModulusChainLink
import com.github.pauldambra.moduluschecker.chain.ModulusResult
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow

class SecondCheckRequiredGate(private val next: ExceptionTwoGate) : ModulusChainLink {

    override fun check(params: ModulusCheckParams): ModulusResult {

        if (exceptionRequiresSecondCheck(params)) {
            return next.check(params)
        }

        return if (params.secondWeightRow != null) {
            next.check(params)
        } else params.modulusResult ?: ModulusResult.PASSES

    }

    private fun exceptionRequiresSecondCheck(params: ModulusCheckParams) =
      rowExceptionRequiresSecondCheck(params.firstWeightRow) || rowExceptionRequiresSecondCheck(params.secondWeightRow)

    private fun rowExceptionRequiresSecondCheck(row: WeightRow?) =
      ModulusResult.exceptionsThatRequireSecondCheck.contains(row?.exception ?: -1)
}
